fedspendingtransparency/data-act-broker-web-app
# DEV-3660 rules filter

**High level description:**
Adds the autocomplete rules filter to the dashboard page.

**Technical details:**
- Reuses multiple components from USAspending, including `Autocomplete` and `ShownValue`
- This filter is disabled until a File is selected. The chosen file is passed as a parameter to the `get_rule_labels` API call 
- Filtering the results by search string is performed on the frontend 
- Adds [FontAwesome icons](https://fontawesome.com/icons?d=gallery) as a dependency

**Link to JIRA Ticket:**
[DEV-3660](https://federal-spending-transparency.atlassian.net/browse/DEV-3660) (rules filter) & [DEV-3876](https://federal-spending-transparency.atlassian.net/browse/DEV-3876) (FontAwesome)

**Mockup**
https://bahdigital.invisionapp.com/share/AXIABX6G347#/screens/296024131

**Testing**

1. On initial page load, hover over the rules text input with no file selected. A warning tooltip should appear to the right of the filter. 
2. Select a file. A loading icon may appear in the rules filter while an API request is made to retrieve the rules corresponding to that file. Click in the rules text input area. The list of rules should appear in a dropdown. 
3. Start typing. The dropdown list should narrow to matching results (case-insensitive). If there are no matching results, an error message should appear below the text input area. 
4. Select a rule from the list. The list should be navigable by keyboard (up and down arrows, enter to select). Once selected, the dropdown should close and a staged filter tag for the selected rule should appear below the text input area. 
5. Repeat step 4. Multiple rules can be selected at a time.
6. Click a filter tag to remove it from the stated filters. 
7. Start typing in the text input area. Without making a selection from the dropdown, click outside the filter. The dropdown should close and the input value should clear. 
 

The following are ALL required for the PR to be merged:
- [x] #1171 merged, change base to `dev`
- [x] Frontend review completed
- [x] Design review completed
- [x] Scheduled Demo including Design/Testing/Front-end OR Provided Instructions for Local Testing above and in JIRA
- [x] Verified cross-browser compatibility
- N/A All componentWillReceiveProps, componentWillMount, and componentWillUpdate in relevant child/parent components/containers replaced with [future compatible life-cycle methods](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html) per [JIRA item 3339](https://federal-spending-transparency.atlassian.net/browse/DEV-3339)
- [x] Link to this PR in JIRA ticket
